source/plugin/placeholders/creatingtokens.rst

```diff
@@ -39,9 +42,154 @@ This is the minimum as specified by Sponge. If another plugin replaces the :java
+
+  * If the string exists and matches a valid :javadoc:`Currency` and the player has an :javadoc:`Account` in that 
+    currency, returns the balance of that account.
+  * If the string exists and does not match a valid :javadoc:`Currency`, or the player does not have an 
+    :javadoc:`Account` in that currency, returns an empty :javadoc:`Text`.
+  * If the string does not exist, if the player has an :javadoc:`Account` in the default :javadoc:`Currency`, return
+    that balance, else return an empty :javadoc:`Text`.
+
+.. code-block:: java
+
+    public class PlayerBalancePlaceholder implements PlaceholderParser {
+        
+        @Override
+        public String getId() {
+            return "spongedocs:balance"
+        }
+
+        @Override
+        public String getName() {
+            return "Balance Placeholder"
+        }
+
+        @Override
+        public Text parse(PlaceholderText placeholderText) {
+            // First: check the EconomyService is available
+            Optional<EconomyService> optionalEconService = Sponge.getServiceManager()
+                .provide(EconomyService.class);
+            if (optionalEconService.isPresent()) {
+                EconomyService econService = optionalEconService.get();
+
+                // This placeholder only applies to players, so check that
+                Optional<Player> player = placeholderText.getAssociatedReceiver()
+                    .map(x -> x instanceof Player ? (Player) x : null);
+                if (player.isPresent()) {
+
+                    // Check the argument string, try to get a valid currency out of it
+                    Currency currency;
+                    if (placeholderText.getArgumentString().isPresent()) {
+                        // If we have an argument string, for this placeholder, we assume the 
+                        // entire string is the argument - so is it a valid Currency ID?
+                        Optional<Currency> co = Sponge.getRegistry()
+                            .getType(Currency.class, placeholderText.getArgumentString().get());
+                        if (!co.isPresent()) {
+                            // If not, then the placeholder is invalid and should return an empty text.
+                            return Text.EMPTY;
+                        }
+                        // It is, so this is what we use
+                        currency = co.get();
+                    } else {
+                        currency = econService.getDefaultCurrency();
+                    }
+
+                    // Finally, check the account exists, and if so, return the value
+                    Optional<UniqueAccount> account = econService
+                        .getOrCreateAccount(player.get().getUniqueId());
+                    return account.map(x -> Text.of(x.getBalance(currency))).orElse(Text.EMPTY);
+                }
+            }
+
+            return Text.EMPTY;
+        }
+    }
```

Comment:
Not sure about this example. It requires a lot of comments in order to be easily understandable.
Have you considered splitting it into multiple parts that can have a translation each?

**Example:**

> ... Introduction ...
>
> - `String getId() { ... }`
> - `String getName() { ... }`
> ````java
> Text parse(PlaceholderText) {
>     Optional<EconomyService> optionalEconService = ...
>     if(!optionalEconService.isPresent()) {
>         return Text.EMPTY;
>     }
>     EconomyService econService = optionalEconService.get();
>
>     Optional<UniqueAccount> optionalAccount = placeholderText.getAssociatedReceiver()
>           .map(recv -> getAccount(econService, recv));
>     if (!optionalAccount.isPresent()) {
>         return Text.EMPTY;
>     }
>     UniqueAccount account = optionalAccount.get();
>
>     Currency currency = placeholderText.getArgumentString()
>           .map(arg -> getCurrency(econService, arg))
>           .orElseGet(econService::getDefaultCurrency);
>     ...
> ````
>
> You can derive the `Account` like this:
>
> - `Optional<Account> getAccount(EconomyService, MessageReceiver) { ... }`
>
> And use the argument string to determine the currency like this:
>
> - `Currency getCurrency(EconomyService, String) { ... }`
>
> But you could also define yourself a separator character (such as `|`) to split the argumentString into more parameters, e.g. adding a number format, turning currency symbols on, or anything you need for your parser.

@Inscrutable What do you think?

Reply:
There are bullet points above that explain the basic workflow. I don't think it's appropriate to describe the economy service in more detail on these pages. I think the bullet points at the top are enough to explain the workflow. For example:

```
You can derive the Account like this:

*    Optional<Account> getAccount(EconomyService, MessageReceiver) { ... }

And use the argument string to determine the currency like this:

    Currency getCurrency(EconomyService, String) { ... }
```

That really isn't appropriate to talk about here and signposts should be made to the economy service docs instead because it detracts from the topic of this document. I will add _that_ in.

I don't really care for the structure of the code one way or another - but it _is_ meant to be a complex example to show different parts of the system. I see the value for the way you've done it though - it does make it easier to see each "section".

I have also previously been asked to put comments into the code. I know they aren't translatable but I still see value of them being there in addition to the bullet points.

I could add a note somewhere above that the string is freeform and can be used however the placeholder wishes, including for multiple arguments, but I don't really want to prescribe how you'd do that.

Reply:
*Musing* I wonder if the more complex example might be appropriate for the Sponge Plugin Cookbook instead - which has been sadly neglected for some time, I believe. When discussion starts to reach its tentacles into other parts of the API, it can be difficult to decide where it fits in the docs, and how to classify it. If it's an important example, perhaps it justifies a new page entirely, rather than dissect it for classification purposes - and link where appropriate. 

As for comments in code blocks, do what you have to do - but try to keep it short and simple, and expand in the text blocks if required.

Reply:
I went for a much simpler example.